components/dada-validate/src/validate/validator.rs

```diff
@@ -520,12 +543,26 @@ impl<'me> Validator<'me> {
                 self.add(validated::ExprData::Seq(validated_exprs), expr)
             }
             syntax::ExprData::Return(with_value) => {
-                if let Some(return_expr) = with_value {
-                    let validated_return_expr = self.validate_expr(*return_expr);
-                    return self.add(validated::ExprData::Return(validated_return_expr), expr);
+                match (self.code.return_type, with_value) {
+                    (Some(return_type), None) => {
+                        dada_ir::error!(self.span(expr), "must return something")
+                            .primary_label("must return something")
+                            .secondary_label(return_type, "because return type is not unit")
+                            .emit(self.db);
+                    }
+                    (None, Some(return_expr)) => {
+                        dada_ir::error!(self.span(*return_expr), "cannot return a value")
+                            .primary_label("cannot return a value")
+                            .emit(self.db);
```

Comment:
```suggestion
                        dada_ir::error!(self.span(*return_expr), "this function doesn't return anything")
                            .primary_label("this function can only `return`, without a value")
                            .emit(self.db);
```

Hmm, it doesn't have to be this PR, but I think it'd be good to show the user *where* the `->` *would* go. For this PR, can we add a label underlining the name and saying something like "because this function doesn't have a `->`"?

Reply:
I'm thinking that can implement it by changing `return_type` to be `FileSpan` which would be the span of `->` or where it should go if it doesn't have it, and separately store a flag like `returns_value`. But I'm not sure it `Code` is the best place to store that information.

Reply:
Maybe a cleaner way of doing this is to have:
```rust
pub enum ReturnType {
    Value(FileSpan), // in the future this could be TokenTree?
    Unit(FileSpan),
}
```

Although seeing that `effect` in `Code` doesn't store `FileSpan` and requires using https://github.com/dada-lang/dada/blob/032f31282da9a9d73b095b8e1784fe99b5407b75/components/dada-validate/src/validate/validator.rs#L34 to be able to later get the span, I'm wondering if there's a reason to avoid storing `FileSpan` or `Filename` in `Code`.

Reply:
I do try to avoid storing `FileSpan` because I want to get more incremental re-use. A good idea when you have something "spanned" is to create an entity for it -- 

e.g.

```rust
salsa::entity2! {
    entity ReturnType in crate::Jar {
        returns_value: bool,
        span: FileSpan,
    }
}
```

Not sure about using a `bool`, but you get the idea.

Reply:
I have created the `ReturnType` entity and `ReturnTypeKind` enum with `Value` or `Unit`, which is a bit of a mouthful but better than a `bool`.